tile/codegen/cstr_reduction.cc

```diff
@@ -15,98 +21,206 @@ using namespace math;    // NOLINT
+      //     min(-rest)
+      //     = -(max(rest))
+      //     = -(max(contraction) - max(idx))
+      //     = max(idx) - max(contraction)
+      //     = ((coeff * range(idx)) - 1) - max(contraction)
+      //     = ((coeff * range(idx)) - 1) - (lim - 1)
+      //     = (coeff * range(idx)) - lim
+      //
+      // So putting it together, the feasible region looks like:
+      //     idx >= ((coeff * range(idx)) - lim_value) / coeff  (rounding upwards)
+      //     idx >= range(idx) - (lim_value / coeff)            (rounding upwards)
+      //
+      // ... and the range [0, range(idx) - (lim_value / coeff)) is infeasible.
+      //
+      // Define:
+      //     shift = range(idx) - RoundUp(lim_value, coeff)
+      //
+      // If shift > 0, we can rewrite all uses of the index, eliding the [0..shift) range that is being
+      // skipped over by this constraint:
+      //     idx' = idx + shift
+      //     range'(idx) = range(idx) - shift
+      //
+      // Note that rewriting the use of the index within the current constraint will also affect the current
+      // min_value and lim_value, by adding the shift to each.
+      //
+      // Also, note that reducing the range of the index may cause other constraints (potentially
+      // already-processed contraints) to become redundant -- which is why we perform the redundant constraint
+      // check after squeezing indicies across all constraints.
+      //
+      // TODO: Consider removing the index entirely when when shift == range(idx), since the index will only
+      //       ever take on the value of zero.
+      if (0 < coeff) {
+        std::int64_t shift = static_cast<int64_t>(idx->range) - RoundUp(lim_value, coeff);
+        if (0 < shift) {
+          IVLOG(4, "Shifting " << kvp.first << " by " << shift);
+          idx->range -= shift;
+          std::string idx_name = kvp.first;
+          Affine new_idx{idx_name};
+          new_idx += shift;
+          for (auto& con : block->constraints) {
+            con.substitute(idx_name, new_idx);
+          }
+          for (auto& dev : block->location.devs) {
+            for (auto& unit : dev.units) {
+              unit.substitute(idx_name, new_idx);
+            }
+          }
+          for (auto& ref : block->refs) {
+            for (auto& access : ref.mut().access) {
+              access.substitute(idx_name, new_idx);
+            }
+            if (ref.cache_unit) {
+              ref.mut().cache_unit->substitute(idx_name, new_idx);
+            }
+          }
+          for (auto& stmt : block->stmts) {
+            if (auto sub = Block::Downcast(stmt)) {
+              for (auto& sidx : block->idxs) {
+                sidx.affine.substitute(idx_name, new_idx);
+              }
+            } else if (auto li = LoadIndex::Downcast(stmt)) {
+              li->from.substitute(idx_name, new_idx);
+            }
+          }
+          min_value += (shift * coeff);
+          lim_value += (shift * coeff);
```

Comment:
I don't think that performing a shift changes `lim_value`.

Reply:
I think you're right -- subtracting the shift from the index's range and then adding it to the uses of the index should cancel out changes to the max and lim values that can be reached by the constraint equation.  Done.